Fix #20268: Minor optimisation in `\yii\helpers\BaseArrayHelper::map`

chriscpty opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | #20268 

Information re: what this does is in the issue.

In the process, I also improved the ArrayHelper::map() tests a bit to also test cases where callbacks are passed in for the three parameters.

chriscpty committed:
Optimise ArrayHelper::map

+ use array_column for faster mapping if no callbacks or grouping are required

chriscpty committed:
Optimise ArrayHelper::map

+ changelog

chriscpty committed:
improve ArrayHelper::map tests

+ also test with callbacks

rob006 commented:
I'm afraid this will change behavior for keys with dots, for example`ArrayHelper::map($array, 'model1.id', 'model2.name')`. Old implementation tried to fetch value from `$arrayItem['model1']['id']`, new one will always use `$arrayItem['model1.id']`.

rhertogh commented on framework/helpers/BaseArrayHelper.php:
```diff
@@ -595,6 +595,9 @@ public static function getColumn($array, $name, $keepKeys = true)
      */
     public static function map($array, $from, $to, $group = null)
     {
+        if (is_string($from) && is_string($to) && $group === null) {
```
Please also add a check that both `$from` and `$to` don't contain a dot (`.`).
The `ArrayHelper::getValue()` function supports getting a value by a "path" string.
Without this check the following code would break:
```php
$array = [
    [
        'prop' => [
            'a' => 'foo',
            'b' => 'bar',
        ],
    ],
];

ArrayHelper::map($array, 'prop.a', 'prop.b');
```

samdark replied:
Yes. It is necessary. @chriscpty would you please redo test with such a check?

chriscpty replied:
performance with the added check:

```
before: 1729233314.6667
after: 1729233315.4382
elapsed: 0.77146291732788
before: 1729233315.4382
after: 1729233316.245
elapsed: 0.80677795410156
before: 1729233316.245
after: 1729233317.0443
elapsed: 0.79932498931885
before: 1729233317.0443
after: 1729233317.8304
elapsed: 0.78613996505737
before: 1729233317.8305
after: 1729233318.6256
elapsed: 0.79515385627747
bench size: 18000
```

I'll add the fix (and related tests) in a second

chriscpty committed:
fix ArrayHelper::map() for path strings

! don't use array_column if ArrayHelper strings are paths
+ add according tests
+ add BaseStringHelper::contains() helper function and according tests

chriscpty commented:
I also added a `BaseStringHelper::contains()` function as php7 does not have `str_contains`. This will become redundant once the minimum supported php version is raised to php8.0 or higher.

chriscpty committed:
fix ArrayHelper::map() for path strings

o code style fixes

rob006 commented:
> I also added a `BaseStringHelper::contains()`

I'm not sure if it is worth to polyfill function, that exists in PHP since 4 years. This method is already redundant for \~70% of framework users.

chriscpty commented:
> This method is already redundant for ~70% of framework users

if that number is true, that'd mean not including the polyfill would be a breaking change for 30% of users. Yii2's minimum supported php version is still 7.3 as far as I'm aware & I don't want to introduce breaking changes for a minor optimisation.

Whether yii2 should bother supporting PHP versions that are EOL is a separate question that is up to the maintainers to answer.

rob006 commented:
> if that number is true, that'd mean not including the polyfill would be a breaking change for 30% of users.

In what scenario not adding **new** function would be a BC break? 70% of framework installations is for PHP8 - in that case developer could use `str_contains()` directly, the rest could stick to `strpos()` or use actual polyfill from symfony. There is no need to add `BaseStringHelper::contains()` in order to implement this minor optimization for `ArrayHelper::map()`.

chriscpty commented:
so in other words, you'd rather have it as a construct like
```php
if (function_exists('str_contains') ? str_contains([...]) : (mb_strpos([...]) !== -1))
```
?

I originally considered doing that, but figured it'd make the condition for the optimisation ugly to read.

rob006 commented:
My suggestion is:
1. Do not add `BaseStringHelper::contains()`.
2. in `Arrayhelper::map()` use `strpos($to, '.') === false` do detect if key contains dot.

If we really want some kind `contains()` helper in `StringHelper`, then IMO it should provide something that `str_contains()` do not provide. For example proper multibyte support - in that case it should consistently use `mb_*`functions and should not fallback to `str_contains()`. But this is out of scope of this PR.

chriscpty committed:
use strpos rather than polyfill

! use strpos rather than polyfill
- remove BaseStringHelper::contains

chriscpty commented:
alright, done :)

samdark commented:
Thank you!

lesha724 commented:
Hі, after these changes `ArrayHelper::map` returns a different result for:

```php
$array = [
    [
         'id' => '1',
         'value' => 'value1',
    ],
    [
         'id' => '2',
    ],
    [
         'id' => '3',
         'value' => 'value3',
    ],
];

ArrayHelper::map($array, 'id', 'value');
```

samdark commented:
What does it return?

lesha724 commented:
Old version returns:

```php
[
   '1' => 'value1',
   '2' => null,
   '3' => 'value3'
]
```
New version returns:

```php
[
   '1' => 'value1',
   '3' => 'value3'
]
```

marcovtwout commented:
This optimization breaks cases that don't pass a regular array but something like a `Traversable` or `Iterator` (for example with Query Builder `->each()`:

```php
return ArrayHelper::map(User::find()->each(), 'id', 'username');
```

Granted, the docblock states the parameter type strictly as `array` but in practice array-like collections are supported on this class.
